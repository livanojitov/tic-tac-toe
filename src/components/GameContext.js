import React, { useState, createContext, useEffect } from 'react';
export const GameContext = createContext();

const GameContextProvider = (props) => {

  const [games, setGames] = useState([]);

  useEffect(() => {
    if (typeof(Storage) !== "undefined") {
      if (games.length){
        localStorage.tictactoe = JSON.stringify(games);
      }else{
        if (GameContextProvider.delete === 1){
          GameContextProvider.delete = 0;
          localStorage.removeItem('tictactoe');
        }else{
          if (localStorage.tictactoe){
            setGames(JSON.parse(localStorage.tictactoe));
          }
        }
      }
    }
  },[games]);

  const addGame = (game) => {
    game.id = Math.floor(Math.random() * 10000);
    setGames([...games, game]);
  };

  const removeGame = (id) => {
    if (games.length > 0){
      GameContextProvider.delete = 1;
      id = id * 1;
      setGames(games.filter(game => game.id !== id));   
    } 
  }

  const getGame = (id) => {
    if (games.length > 0){
      id = id * 1;
      return games.filter((game) => game.id === id)[0];
    }else{
      return false;
    }    
  }

  return (
    <GameContext.Provider value={{ games, addGame, removeGame, getGame }}>
      {props.children}
    </GameContext.Provider>
  );

}
 
export default GameContextProvider;