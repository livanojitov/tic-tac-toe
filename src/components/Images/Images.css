.images{
  margin: 10px 0;
  font-weight: bold;
}

.image-token{
  width: 120px;
  height: 120px;
  vertical-align: middle;
}

.user-image, .computer-image{
  display: inline-block;
}

.user-image{
  width: 40%;
}

.computer-image{
  width: 60%;
} 

.refresh{
  width: 30px;
  height: 30px;  
  margin-left: 5px;
  position: relative;
  top: 12px;
  cursor: pointer;
}

@media screen and (max-width: 400px) {
  .image-token{
    width: 40px;
    height: 40px;
  }  
}