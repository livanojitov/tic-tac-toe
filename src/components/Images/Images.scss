.images {
  margin: 10px 0;
  font-weight: bold;

  .user-image,
  .computer-image {
    display: inline-block;
    width: 50%;
  }

  span {
    width: 37%;
    display: inline-block;
  }

  .image-token {
    width: 40%;
    vertical-align: middle;
  }

  .refresh {
    width: 13%;
    margin: 0 1%;
    position: relative;
    vertical-align: middle;
    cursor: pointer;
  }
}

.historical.images{
  .image-token {
    width: 54%;
  }  
}
