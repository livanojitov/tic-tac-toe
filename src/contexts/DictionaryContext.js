import React, { createContext } from "react";
import useFetchData from "../hooks/useFetchData";
import useLanguage from "../hooks/useLanguage";
import * as constants from "../Constants";

export const DictionaryContext = createContext();

const DictionaryContextProvider = ({ children }) => {
  let { data: dictionary, error } = useFetchData("dictionary");

  const [language, changeLanguageHandler] = useLanguage(constants.ENGLISH);

  if (error) {
    return (
      <div className="error" data-testid="error">
        {error}
      </div>
    );
  }

  dictionary = dictionary[language];

  return (
    <DictionaryContext.Provider value={{ dictionary, changeLanguageHandler }}>
      {children}
    </DictionaryContext.Provider>
  );
};

export default DictionaryContextProvider;
