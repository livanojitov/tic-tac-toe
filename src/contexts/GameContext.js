import React, { createContext, useContext, useEffect, useReducer } from "react";
import { CategoryContext } from "./CategoryContext";
import { gameReducer } from "../store/game/gameReducer";
import { USER, RANDOM, IMAGES_PER_CATEGORY } from "../Constants";
import { changeCategory } from "../store/game/gameActions";

export const GameContext = createContext();

const initialState = {
  category: {
    id: 5,
    folder: "soccer",
    name: "SOCCER_PLAYERS",
  },

  images: {
    user: Math.floor(Math.random() * IMAGES_PER_CATEGORY),
    computer: Math.floor(Math.random() * IMAGES_PER_CATEGORY),
  },

  first: USER,

  level: RANDOM,

  disabled: true,

  over: false,
};

const GameContextProvider = ({ children }) => {
  const [game, dispatch] = useReducer(gameReducer, initialState);
  const { categories } = useContext(CategoryContext);

  useEffect(() => {
    dispatch(
      changeCategory(categories[Math.floor(Math.random() * categories.length)])
    );
  }, [categories]);

  return (
    <GameContext.Provider value={{ game, dispatch }}>
      {children}
    </GameContext.Provider>
  );
};

export default GameContextProvider;
