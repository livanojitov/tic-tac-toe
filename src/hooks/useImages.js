import { useState, useCallback } from "react";
import { IMAGES_PER_CATEGORY } from "../Constants";

function useImages(user = 0, computer = 1) {
  const [images, setImages] = useState({ user, computer });

  const randomImage = useCallback(() => {
    let image = Math.floor(Math.random() * IMAGES_PER_CATEGORY);
    while (image === images.computer || image === images.user) {
      image = Math.floor(Math.random() * IMAGES_PER_CATEGORY);
    }
    return image;
  }, [images.computer, images.user]);

  const refreshImage = useCallback(
    (e) => {
      setImages({
        ...images,
        [e.target.id]: randomImage(),
      });
    },
    [images, randomImage]
  );

  if (images.user === images.computer) {
    setImages({
      ...images,
      user: randomImage(),
    });
  }

  return { images, refreshImage };
}

export default useImages;
