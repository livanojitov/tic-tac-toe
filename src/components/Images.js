import React, { useContext, useState } from 'react';
import { HistoryContext }              from '../contexts/HistoryContext';
import { LanguageContext }             from '../contexts/LanguageContext';
import {COMPUTER, USER}                from './Constants';
import { DictionaryContext }           from '../contexts/DictionaryContext';

const Images = ({folder, count, imageUser:imageUser1, imageComputer:imageComputer1, onImageChange}) => {
  const { getLanguage } = useContext(LanguageContext);
  const language = getLanguage();  
  const { history } = useContext(HistoryContext);
  const { DICTIONARY } = useContext(DictionaryContext);
  const [ imageUser, setImageUser ] = useState(imageUser1);  
  const [ imageComputer, setImageComputer ] = useState(imageComputer1);
  
  const randomImage = () => {
    let image = Math.floor(Math.random() * count);
    while ( image === imageComputer || image === imageUser ){
      image = Math.floor(Math.random() * count);
    }
    return image;
  }
  
  const refreshUser = () => {
    let image = randomImage();
    setImageUser(image);
    onImageChange({
      imageUser : image
    });
  }

  const refreshComputer = () => {
    let image = randomImage();  
    setImageComputer(image);
    onImageChange({
      imageComputer : image
    });
  }

  return (
      <div className={ history ? "images history" : "images"}>

        <div className='user-image'>

          <span>{DICTIONARY && DICTIONARY[language].YOU}: </span>

          <img className="imageToken" 
               src={`./images/${folder}/${imageUser}.jpg`} 
               alt="user" 
               title={ !history ? DICTIONARY && DICTIONARY[language].USER_IMAGE_TOOLTIP : "" }/>
          
          { !history && ( 
            <img className='refresh' 
                 src={`./images/refresh.png`} 
                 alt="refresh" 
                 onClick={()=>refreshUser(USER)}
                 title={DICTIONARY && DICTIONARY[language].REFRESH_TOOLTIP} />)}
        </div> 
        
        <div className='computer-image'>

          <span>{DICTIONARY && DICTIONARY[language].OPPONENT}: </span>

          <img className="imageToken"
               src={`./images/${folder}/${imageComputer}.jpg`}
               alt="computer"
               title={ !history ? DICTIONARY && DICTIONARY[language].COMPUTER_IMAGE_TOOLTIP : "" }/>

          { !history && ( 
            <img className='refresh'
                 src={`./images/refresh.png`} 
                 alt="refresh" 
                 onClick={()=>refreshComputer(COMPUTER)}
                 title={DICTIONARY && DICTIONARY[language].REFRESH_TOOLTIP} />)}

        </div>
        
      </div>
  )
  
}

export default Images;