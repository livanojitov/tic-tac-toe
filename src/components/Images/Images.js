import React, { useContext, useEffect, useMemo } from "react";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import { GameContext } from "../../contexts/GameContext";
import { changeImages } from "../../store/game/gameActions";
import useImages from "../../hooks/useImages";
import "./Images.scss";

const Images = () => {
  const { dictionary } = useContext(DictionaryContext);
  const { game, dispatch } = useContext(GameContext);
  const folder = game.category.folder;

  const url =
    "https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images";
  const { images, refreshImage } = useImages(
    game.images.user,
    game.images.computer
  );
  const { user: imageUser, computer: imageComputer } = images;

  useEffect(() => {
    dispatch(
      changeImages({
        user: imageUser,
        computer: imageComputer,
      })
    );
  }, [imageUser, imageComputer, dispatch]);

  return useMemo(() => {
    return dictionary && folder ? (
      <div className="images" data-testid="images">
        <div className="user-image">
          <span>{dictionary.YOU}: </span>
          <img
            className="image-token"
            src={`${url}%2F${folder}%2F${imageUser}.jpg?alt=media`}
            alt="user"
            title={dictionary.USER_IMAGE_TOOLTIP}
          />
          <img
            className="refresh"
            id="user"
            src={`${url}%2Frefresh.png?alt=media`}
            alt="refresh"
            onClick={(e) => refreshImage(e)}
            title={dictionary.REFRESH_TOOLTIP}
            data-testid="user-image"
          />
        </div>

        <div className="computer-image">
          <span>{dictionary.OPPONENT}: </span>
          <img
            className="image-token"
            src={`${url}%2F${folder}%2F${imageComputer}.jpg?alt=media`}
            alt="computer"
            title={dictionary.COMPUTER_IMAGE_TOOLTIP}
          />
          <img
            className="refresh"
            id="computer"
            src={`${url}%2Frefresh.png?alt=media`}
            alt="refresh"
            onClick={(e) => refreshImage(e)}
            title={dictionary.REFRESH_TOOLTIP}
            data-testid="computer-image"
          />
        </div>
      </div>
    ) : null;
  }, [dictionary, imageUser, imageComputer, refreshImage, folder]);
};

export default Images;
