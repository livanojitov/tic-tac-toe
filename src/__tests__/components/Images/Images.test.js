import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from '@testing-library/user-event'
import "@testing-library/jest-dom/extend-expect";
import { DictionaryContext } from "../../../contexts/DictionaryContext";
import { GameContext } from "../../../contexts/GameContext";
import Images from "../../../components/Images/Images";
import useImages from "../../../hooks/useImages";

jest.mock("../../../hooks/useImages");

describe("Images", () => {
  let dictionary;
  let game;
  let dispatch;
  let error;
  let refreshImageHandler;

  beforeEach(() => {
    dictionary = {
      YOU: "You",
      USER_IMAGE_TOOLTIP: "You will play with this image.",
      REFRESH_TOOLTIP: "Click to play with a different image.",
      OPPONENT: "PC",
      COMPUTER_IMAGE_TOOLTIP: "The computer will play with this image.",
    };

    game = {
      category: {
        id: 0,
        folder: "habana",
        name: "HAVANA_LANDMARKS",
      },
      images: {
        user: 0,
        computer: 1,
      },
    };

    dispatch = jest.fn();

    error = "";

    refreshImageHandler = jest.fn();
  });

  const renderImages = () => {
    useImages.mockReturnValue({
      images: { user: 0, computer: 1 },
      refreshImage: refreshImageHandler,
    });

    return render(
      <DictionaryContext.Provider value={{ dictionary, error }}>
        <GameContext.Provider value={{ game, dispatch }}>
          <Images />
        </GameContext.Provider>
      </DictionaryContext.Provider>
    );
  };

  it("renders the Image component correctly", () => {
    const { asFragment } = renderImages();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should not display the images if dictionary hasn't been fetched yet", () => {
    dictionary = "";

    renderImages();

    expect(screen.queryByTestId("images")).toBeNull();
  });

  it("should display the images if dictionary has already been fetched", () => {
    renderImages();

    expect(screen.getByTestId("images")).not.toBeNull();
  });

  it("should call the refresh image handler when clicking on the user image", () => {
    renderImages();

    userEvent.click(screen.getByTestId("user-image"));

    expect(refreshImageHandler).toHaveBeenCalled();
  });

  it("should call the refresh image handler when clicking on the computer image", () => {
    renderImages();

    userEvent.click(screen.getByTestId("computer-image"));

    expect(refreshImageHandler).toHaveBeenCalled();
    expect(refreshImageHandler).toHaveBeenCalled();
  });
});
