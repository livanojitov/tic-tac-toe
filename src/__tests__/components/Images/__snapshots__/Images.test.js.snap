// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Images renders the Image component correctly 1`] = `
<DocumentFragment>
  <div
    class="images"
    data-testid="images"
  >
    <div
      class="user-image"
    >
      <span>
        You: 
      </span>
      <img
        alt="user"
        class="image-token"
        src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F0.jpg?alt=media"
        title="You will play with this image."
      />
      <img
        alt="refresh"
        class="refresh"
        data-testid="user-image"
        id="user"
        src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Frefresh.png?alt=media"
        title="Click to play with a different image."
      />
    </div>
    <div
      class="computer-image"
    >
      <span>
        PC: 
      </span>
      <img
        alt="computer"
        class="image-token"
        src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F1.jpg?alt=media"
        title="The computer will play with this image."
      />
      <img
        alt="refresh"
        class="refresh"
        data-testid="computer-image"
        id="computer"
        src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Frefresh.png?alt=media"
        title="Click to play with a different image."
      />
    </div>
  </div>
</DocumentFragment>
`;
